packages/ses/src/tame-faux-data-properties.js

```diff
@@ -0,0 +1,124 @@
+import {
+  getOwnPropertyDescriptor,
+  apply,
+  defineProperty,
+  toStringTagSymbol,
+} from './commons.js';
+
+/**
+ * Exported for convenience of unit testing. Harmless, but not expected
+ * to be useful by itself.
+ *
+ * @param {any} obj
+ * @param {string|symbol} prop
+ * @param {any} expectedValue
+ * @returns {boolean}
+ */
+export const tameFauxDataProperty = (obj, prop, expectedValue) => {
+  if (obj === undefined) {
+    return false;
+  }
+  const desc = getOwnPropertyDescriptor(obj, prop);
+  if (!desc || 'value' in desc) {
+    return false;
+  }
+  const { get, set } = desc;
+  if (typeof get !== 'function') {
+    return false;
+  }
+  if (typeof set !== 'function') {
+    return false;
+  }
+  if (apply(get, obj, []) !== expectedValue) {
+    return false;
+  }
+  const testValue = 'Seems to be a setter';
+  const subject1 = { __proto__: null };
+  apply(set, subject1, [testValue]);
+  if (subject1[prop] !== testValue) {
+    return false;
+  }
+  const subject2 = { __proto__: obj };
+  apply(set, subject2, [testValue]);
+  if (subject2[prop] !== testValue) {
+    return false;
+  }
+
+  // TODO We would like to test that the setter throws when attempting to
+  // set the target object. However, since we don't already know what it
+  // does, the following test could cause a persistent side effect.
```

Comment:
So I just read the proposal steps and I'm actually somewhat surprised at the specified behavior for [SetterThatIgnoresPrototypeProperties](https://tc39.es/proposal-iterator-helpers/#sec-SetterThatIgnoresPrototypeProperties). I would expect that if `this === home` it would attempt a define property. Arguably `Iterator.prototype.contstructor = Iterator` should never fail (but also possibly not change the property), and direct assignment to another value shouldn't fail either if `constructor` is configurable.

I know the spec there is exactly how we implemented our property override taming, but we assumed lockdown (which may be a mistake?), but the spec shouldn't assume that the intrinsic prototypes are frozen.

Reply:
Given the above, it might be safe to test the setter with the expected value, and restore the throw test? If the setter is not compliant, worst case it overrides with the expected value as a side effect, which should maintain semantics. FWIW, the other setter tests might have side effects too if non compliant. And if we want to be pedantic, calling the getter might have side effects too.

Reply:
Regarding the behavior when assigning to a non-writable property, I forgot that it throws for same value (unlike define to same desc, such a weird discrepancy). It doesn't change my surprise that the spec implements `constructor` as if it was a non-writable property. It also doesn't change my acceptance that we could test the throwing behavior with the expected value. It just changes my understanding of what the spec should do here.

Reply:
> would attempt a define property

define the property to what? Remember, by assumption, it is an accessor property. If it redefined the property to the data property that would normally be the result of a successful assignment to a data property, then that property, if subsequently made non-writable, will trigger the override mistake. I don't think attempting to define a property on home is an option here.

> direct assignment to another value shouldn't fail either if `constructor` is configurable

Neither assignment failure nor override mistake failures depend at all on configurability. They both depend only on writability. However, as an accessor property emulating a data property storing only the emulated value off to the side, it has no state to track whether it should emulate writable or non-writable. Extending the spec to burden the emulation with tracking emulated writability is, I suppose, possible. But I think it is great we did not do that.

> FWIW ... side effects too.

Good point! I'm going to restore the setter test. But I really like your suggestion to invoke the setter with the already-validated expected value. That way, if it fails by performing the most obvious emulated assignment, it will not actually have changed any observable state. Or so one can hope.

> It just changes my understanding of what the spec should do here

I still think the spec is doing the least bad thing it is possible for the spec to do, given all the constraints. If you still disagree, we should discuss.

Reply:
In light of this thread, I restored these stricter checks but added comments. Resolving.